¡Hola, Christian!

Mi nombre es Tonatiuh Cruz. Me complace revisar tu proyecto hoy.

Al identificar cualquier error inicialmente, simplemente los destacaré. Te animo a localizar y abordar los problemas de forma independiente como parte de tu preparación para un rol como data-scientist. En un entorno profesional, tu líder de equipo seguiría un enfoque similar. Si encuentras la tarea desafiante, proporcionaré una pista más específica en la próxima iteración.

Encontrarás mis comentarios a continuación - **por favor no los muevas, modifiques o elimines**.

Puedes encontrar mis comentarios en cajas verdes, amarillas o rojas como esta:

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Éxito. Todo está hecho correctamente.
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Observaciones. Algunas recomendaciones.
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Necesita corrección. El bloque requiere algunas correcciones. El trabajo no puede ser aceptado con comentarios en rojo.
</div>

Puedes responderme utilizando esto:

<div class="alert alert-block alert-info">


<div class="alert alert-block alert-success">
<b>Review General iteración 1) </b> <a class="tocSkip"></a>
    
Christian, has hecho un excelente trabajo con el desarrollo del proyecto, cada vez más cercas de convertirte en un cientifico de datos. Realizaste la carga de bases, su análisis inicial,el desarrollo de los modelos, el calculo del RMSE además lo replicate para todas la regiones y concluiste con el análisis de ganancias. Solamente te dejo algunos comentarios para complementar el análisis.

Sigue con el excelente trabajo!

In [1]:
from dataclasses import dataclass
from typing import Dict, Tuple, List
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import StandardScaler

In [2]:
df_0 = pd.read_csv('/datasets/geo_data_0.csv')
df_1 = pd.read_csv('/datasets/geo_data_1.csv')
df_2 = pd.read_csv('/datasets/geo_data_2.csv')

In [3]:
df_0.head(5)

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

In [4]:
df_1.head(5)

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

In [5]:
df_2.head(5)

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

In [6]:
features_0 = df_0.drop(['id', 'product'], axis=1)
target_0 = df_0['product']

features_1 = df_1.drop(['id', 'product'], axis=1)
target_1 = df_1['product']

features_2 = df_2.drop(['id', 'product'], axis=1)
target_2 = df_2['product']

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Excelente trabajo con la carga de la información y su exploración inicial.  Solamente te recomiendo un análisis perfecto de los registros duplicados que como aprendiste en cursos pasados sirve para asegurar que nuestros resultados no estarán sesgados

In [7]:
X_train, X_valid, y_train, y_valid = train_test_split(
    features_0, target_0, test_size=0.25, random_state=42
)

model = LinearRegression()
model.fit(X_train, y_train)

preds = model.predict(X_valid)

mse = mean_squared_error(y_valid, preds)
rmse = mse ** 0.5

print("Región 0 - RMSE:", rmse)
print("Media de predicciones:", preds.mean())
print("Media real:", y_valid.mean())

Región 0 - RMSE: 37.75660035026169
Media de predicciones: 92.39879990657768
Media real: 92.32595637084387


In [16]:
def evaluar_region(features, target, region_name):
    # Dividir en train/valid
    X_train, X_valid, y_train, y_valid = train_test_split(
        features, target, test_size=0.25, random_state=42
    )
    
    # Escalar datos
    scaler = StandardScaler()
    X_train_scaled = scaler.fit_transform(X_train)
    X_valid_scaled = scaler.transform(X_valid)
    
    # Entrenar modelo
    model = LinearRegression()
    model.fit(X_train_scaled, y_train)
    
    # Predicciones
    preds = model.predict(X_valid_scaled)
    
    # Métricas
    rmse = mean_squared_error(y_valid, preds) ** 0.5
    r2 = model.score(X_valid_scaled, y_valid)
    
    print(f"Región {region_name} - RMSE: {rmse:.2f}")
    print(f"R²: {r2:.2f}")
    print(f"Media de predicciones: {preds.mean():.2f}")
    print(f"Media real: {y_valid.mean():.2f}")
    print("-" * 50)
    
    return preds, y_valid

In [17]:
features_0 = df_0.drop(['id', 'product'], axis=1)
target_0 = df_0['product']

features_1 = df_1.drop(['id', 'product'], axis=1)
target_1 = df_1['product']

features_2 = df_2.drop(['id', 'product'], axis=1)
target_2 = df_2['product']

In [18]:
preds_0, y_valid_0 = evaluar_region(features_0, target_0, "0")
preds_1, y_valid_1 = evaluar_region(features_1, target_1, "1")
preds_2, y_valid_2 = evaluar_region(features_2, target_2, "2")

Región 0 - RMSE: 37.76
R²: 0.27
Media de predicciones: 92.40
Media real: 92.33
--------------------------------------------------
Región 1 - RMSE: 0.89
R²: 1.00
Media de predicciones: 68.71
Media real: 68.73
--------------------------------------------------
Región 2 - RMSE: 40.15
R²: 0.20
Media de predicciones: 94.77
Media real: 95.15
--------------------------------------------------


<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Christian, implementaste el modelo de regresión lineal de forma excelente para los 3 conjuntos de datos! En este punto ya puedes obervar cuál es el modelo con el mayor $R^2$ y menor RMSE. 
    
Solamente recuerda complementarlo con el escalado de los datos. El escalado es muy útil para mejorar el rendimiento del modelo al asegurar que todas las características tienen el mismo rango.     
</div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Gran trabajo con el ajuste!

In [19]:
PRICE_PER_UNIT = 4500
BUDGET = 100_000_000
TOP_WELLS = 200

def calcular_beneficio_top(preds, y_valid):
    # Obtener índices de los 200 mejores pozos según predicción
    idx_top = np.argsort(preds)[::-1][:TOP_WELLS]
    y_top = y_valid.iloc[idx_top]  # valores reales de estos pozos
    
    # Beneficio
    beneficio = np.sum(y_top) * PRICE_PER_UNIT - BUDGET
    return beneficio, y_top

In [21]:
# Región 0
benef_0, y_top_0 = calcular_beneficio_top(preds_0, y_valid_0)
print(f"Región 0 - Beneficio top 200: ${benef_0:,.0f}")

# Región 1
benef_1, y_top_1 = calcular_beneficio_top(preds_1, y_valid_1)
print(f"Región 1 - Beneficio top 200: ${benef_1:,.0f}")

# Región 2
benef_2, y_top_2 = calcular_beneficio_top(preds_2, y_valid_2)
print(f"Región 2 - Beneficio top 200: ${benef_2:,.0f}")


Región 0 - Beneficio top 200: $33,591,411
Región 1 - Beneficio top 200: $24,150,867
Región 2 - Beneficio top 200: $25,985,718


In [22]:
N_BOOT = 1000
POINTS_STUDIED = 500  # como indica el proyecto

def bootstrap_beneficio(preds_valid, y_valid, n_boot=N_BOOT, sample_size=POINTS_STUDIED, top_wells=TOP_WELLS):
    rng = np.random.default_rng(42)
    profits = []
    
    # Empaquetamos pares predicción/valor real
    pairs = np.column_stack([preds_valid, y_valid.values])
    
    for _ in range(n_boot):
        sample_idx = rng.integers(0, len(pairs), size=sample_size)
        sample = pairs[sample_idx]
        preds_s = sample[:,0]
        y_s = sample[:,1]
        
        # Seleccionamos top pozos
        idx_top = np.argsort(preds_s)[::-1][:top_wells]
        y_top = y_s[idx_top]
        
        # Calculamos beneficio
        profit = np.sum(y_top) * PRICE_PER_UNIT - BUDGET
        profits.append(profit)
    
    profits = np.array(profits)
    mean_profit = profits.mean()
    ci_low, ci_high = np.percentile(profits, [2.5, 97.5])
    risk = np.mean(profits < 0)
    
    return mean_profit, ci_low, ci_high, risk

In [23]:
# Región 0
mean_0, ci_low_0, ci_high_0, risk_0 = bootstrap_beneficio(preds_0, y_valid_0)
print(f"Región 0 - Beneficio promedio: ${mean_0:,.0f}, IC95%: [{ci_low_0:,.0f}, {ci_high_0:,.0f}], Riesgo: {risk_0*100:.2f}%")

# Región 1
mean_1, ci_low_1, ci_high_1, risk_1 = bootstrap_beneficio(preds_1, y_valid_1)
print(f"Región 1 - Beneficio promedio: ${mean_1:,.0f}, IC95%: [{ci_low_1:,.0f}, {ci_high_1:,.0f}], Riesgo: {risk_1*100:.2f}%")

# Región 2
mean_2, ci_low_2, ci_high_2, risk_2 = bootstrap_beneficio(preds_2, y_valid_2)
print(f"Región 2 - Beneficio promedio: ${mean_2:,.0f}, IC95%: [{ci_low_2:,.0f}, {ci_high_2:,.0f}], Riesgo: {risk_2*100:.2f}%")

Región 0 - Beneficio promedio: $4,133,525, IC95%: [-944,616, 9,282,041], Riesgo: 4.50%
Región 1 - Beneficio promedio: $4,576,622, IC95%: [624,837, 8,436,717], Riesgo: 1.50%
Región 2 - Beneficio promedio: $3,952,773, IC95%: [-1,288,119, 8,816,560], Riesgo: 6.80%


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Muy buen trabajo con el calculo de las ganacias para cada una de las regiones. Posteriormente este resultados lo complementas con el análisis de riesgos


Hola! Quisiera en este proyecto presentar mi conclusion como si fuera entrega de un proyecto como si fuera ya de un trabajo, esto es mas un extra para mi, si puedes darme solo una retro de como lo hice no a nivel proyecto de clase si no a nivel proyecto de trabajo, asi que lo intentare a partir de aqui:

Se realizo el modelo de regresion lineal para predecir el volumen de reservas de los pozos de las 3 regiones:

La Region 1 muestra un benefio promedio alto y muy bajo riesgo de perdida 1.5% lo que conviene en la opcion mas segura
La region 0 tiene beneficio de potencial mas alto pero mayor riesgo con 4.5% de perdida debido al error del model es arriesgada
La region 2 tiene un beneficio de promedio decente pero el riesgo es aun mas alto con 6.8% de perdida por lo que tambien es una decision arriesgada

Recomiendo fuertemente la region 1 ya que esta es la region con el valor de riesgo de perdida de 1.5% aun cuando el beneficio es menor ***ademas de que el coeficiente dedeterminacion dio como resultado 1.00 lo cual es literalmente perfecto, el modelo de aprendio la relacion casi perfectamente, lo cual nos garantiza casi nada de error***.

Con mas tiempo y datos adicionales podria mejorar el modelo 0 y 2 utilizando arbol de decision, RandomForest, etc.

Por el momento esta es mi conclusion, quedo atento a cualquier feedback, gracias, saludos